### DISCUSSION TITLE
firefox: cannot use speech-dispatcher (TTS)

### DISCUSSION THREAD

**Discussion by sak96** [netblue30/firejail#5762] [Q&A]
- speech-dispatcher allow **TTS** on firefox.

* **what has been done till now?**
  firefox.local is updated as follows
  ```firejail
  whitelist ${RUNUSER}/speech-dispatcher/*
  ```
* **is it issue with tool which is doing TTS?**
   no if i run tool via command line. it works
    ```bash
   cat "test" |  firejaill --profile=firefox RHVoice-test
    ```
 * **what do you think is the issue?**
   if i run `spd-say` it provide following error:
   ```
    firejail --quiet --profile=firefox spd-say "test"
    Failed to connect to Speech Dispatcher:
    Error: Can't connect to unix socket /run/user/1000/speech-dispatcher/speechd.sock: Connection refused. Autospawn: Autospawn failed. Speech Dispatcher refused to start with error code, stating this as a reason: Error: can't open logging file /var/log/speech-dispatcher//speech-dispatcher.log! Using stdout.
   ```
   my guess is that the  connection refused is the cause of error, But i don't see any change in permission or use both inside and outside of profile

**Comment by rusty-snake** [comment]
> whitelist ${RUNUSER}/speech-dispatcher/*

Whitelist the directory and not the socken. And make sure the directory exists when starting firejail.

**Reply by sak96** [reply]
i just now found the solution. still making sense of what is happening.

Looks like i had to do
```firejail
whitelist ${HOME}/.config/speech-dispatcher/*
read-only ${HOME}/.config/speech-dispatcher/*
# whitelist ${RUNUSER}/speech-dispatcher/*
```
Note: if i un-comment the last line the speech-dispatcher doesn't work.

Maybe if the socket is not there and config is present in the home directory for speech-dispatcher it use some other way of communicating which doesn't involve the socket. And socket communication may be failing due to some thing else (may be auth/some file missing).

**Reply by sak96** [reply]
It looks like that as the socket is not present when i join firefox profie.
~also i removed the libraries required by the rhvoice tts tool.~ This had to be reverted


``firejail
whitelist /usr/share/RHVoice/*
read-only /usr/share/RHVoice/*
```
and it is still working.

Thanks for the help

**Comment by omega3** [comment]
When I use Firefox without firejail speech dispatcher works ok. But when I run it within firejail with local profile there is no sound (although normally there is sound from youtube etc).

```
firejail --private=/home/user/jail/ --profile=/home/user/jail/.config/firejail/firefox.local /usr/lib/firefox/firefox %u


whitelist ${HOME}/.config/speech-dispatcher/*
read-only ${HOME}/.config/speech-dispatcher/*
#whitelist ${RUNUSER}/speech-dispatcher/*
whitelist ${HOME}/.local/share/com.mikeasoft.pied/*
read-only ${HOME}/.local/share/com.mikeasoft.pied/*

noblacklist ${HOME}/.config/speech-dispatcher/*
read-only ${HOME}/.config/speech-dispatcher/*
#noblacklist ${RUNUSER}/speech-dispatcher/*
noblacklist ${HOME}/.local/share/com.mikeasoft.pied/*
read-only ${HOME}/.local/share/com.mikeasoft.pied/* 

```
I just copied [Pied](https://github.com/Elleo/pied) folder and speech-dispatcher folder from normal /home but I changed paths there.

`firejail --quiet --profile=/home/user/jail/.config/firejail/firefox.local spd-say "test"`
produces no sound neither error messages.

The strange thing is that when in firefox in reader mode I press headphone icon I see voices selection so Pied voices are visible as an option but no sound. Any ideas?

**Edition:**

`firejail --quiet --noprofile spd-say "test"`

produces sound, so I guess it is a matter of proper profile.

Pied creates this config:

```
GenericExecuteSynth "if command -v sox > /dev/null; then\
        PROCESS=\'sox -r 22050 -c 1 -b 16 -e signed-integer -t raw - -t wav - tempo $RATE pitch $PITCH norm\'; OUTPUT=\'$PLAY_COMMAND\';\
    elif command -v paplay > /dev/null; then\
        PROCESS=\'cat\'; OUTPUT=\'$PLAY_COMMAND --raw --channels 1 --rate 22050\';\
    else\
        PROCESS=\'cat\'; OUTPUT=\'aplay -t raw -c 1 -r 22050 -f S16_LE\';\
    fi;\
    echo \'$DATA\' | /home/user/.local/share/com.mikeasoft.pied/piper/piper --model /home/user/.local/share/com.mikeasoft.pied/models/en_US-libritts_r-medium.onnx --output_raw | $PROCESS | $OUTPUT;"
GenericRateAdd 1
GenericPitchAdd 1
GenericVolumeAdd 1
GenericRateMultiply 1
GenericPitchMultiply 1000
AddVoice "en-us" "MALE1" "Piper"

```
I just copied it to my custom /home and adjusted path 
${HOME}/.local/share/com.mikeasoft.pied/piper/piper

Even if I give full path to my jailed home
/home/user/jail/.config/speech-dispatcher/
nothing changes.

I wonder if some audio processing problem, sockets or whatever are involved that I should include in profile. I don't know how to track down the cause.

I don't need to necessarily use Pied. I will be grateful for working piper and speech dispatch config files that work with firejail custom profile.
